### DISCUSSION TITLE
tutorials/best_practices/node_alternatives

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#163] [User-contributed notes]
# tutorials/best_practices/node_alternatives

Nodes are cheap to produce, but even they have their limits. A project may have tens of thousands of nodes all doing things. The more complex their behavior though, the larger the strain each one a...

https://docs.godotengine.org/en/stable/tutorials/best_practices/node_alternatives.html

**Comment by buanzo** [comment]
Use Nodes for visual and interaction purposes but check the performance as your node count grows. Replace data-centric nodes with Object or RefCounted where possible.. THen, Resources for data that benefits from serialization, like settings, configurations, or reusable assets, help with performance and keep the scene tree clean.

**Comment by badunius** [comment]
Never use Custom Resources for anything that **isn't** supposed to be shared between the instances. Unless you're going out of your way to make sure they are unique, they won't be. And wen they are, you're no longer benefiting from them being accessible from the editor.

**Reply by Nrosa01** [reply]
What do you mean? Let's say you have a Character resource and this one has an array of CharacterUpgrade, being Upgrade another resource. Each upgrade is a differente resource, unique to the Character and each Upgrade is different. You're still gaining value, the value of being able to edit them from the inspector.

**Reply by badunius** [reply]
I only mean, if you are not careful enough, using resources may lead to an unwanted outcomes. Resource has yo be marked as local, and if you miss a single checkbox, you're up yo some debugging fun.

The were created to be shared between nodes, and this should stay the only use case, everything else is solvable with RefCounted instances.

**Comment by XandaPanda42** [comment]
Word of warning: Don't rely on `RefCounted` and `Resource` if you need to clean up your objects before freeing them. Neither of these classes will successfully run a `NOTIFICATION_PREDELETE` call, as the pointer to the Resource gets freed before the notification runs. 

There's no mention of this anywhere in the documentation that I could find, but an in-Editor memory leak caused by a silently failing tool script destructor is not fun to debug. It used to at least warn us that something was wrong when it tried to run the code, now it just fails with no error.

There's a string of issues dating back to at least [2016](https://github.com/godotengine/godot/issues/6784) talking about this, and still no warning anywhere on the user side (nothing on the `RefCounted` or `Resource` pages, nothing on this page, no red or yellow line in the script editor.) Except the issues page and a source comment that basically said "just don't do that".

If you use the Server singletons, or need to manage your own RIDs or filesystem or other data and want any kind of "destructor" function, you'll need to extend Object, or Node, or bodge together a hacky solution.

The built in classes get freed as expected as far as I can tell, so I think it's limited only to the base classes and scripts that extend them.
